### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by nlukic97**
Changes:
- Updated README
- Moved validation logic into separate middleware called in router
- Added two passport.js strategies called in router after data validation:
    -  JWT for authentication via JWT bearer token. Added this to all routes non-GET routes and the "/login" POST route
    - Local for user login, added to the "/login" POST route
- Added field "companyOwner" to the table "companies" (with the  One-To-Many association between User and Company model)
- Updated /register method to add a new User, Company, and Profile (with mandatory and optional parameters handled)
- Validation that ensures only users who "own" a profile or company can make updates to them. This does not include the delete methods, any logged in user can delete another profile/ company/ user - waiting for further task instructions on this.
- Use of Sequelize transactions when making insert queries for the POST method "/register".

Notes:
- In the ***profiles*** table, the field _user_ cannot be updated after insert.
- In the ***companies*** table, the _companyOwner_ field cannot be updated after insert. Superadmin will be able to, but this is for the next task.
- Have 4 TypeScript errors on included Models on a SELECT database queries (in the first 4 comments of this PR), still searching for solution

### CODE REVIEW

File: ts/controllers/Company.controller.ts [hunk 1]
```diff
@@ -88,78 +64,35 @@ const getCompanyById = async (req:Request, res:Response) =>{
...
+            where:{
+                id: req.params.id
+            },
+            include:User
+        })
         
         if(companyExists === null) throw `A company with the id ${req.params.id} does not exist.`;
+        
+        // If the user who has the company is not the same as the user who is making this request (from passport-jwt)        
+        if(companyExists.User.id !== req.user.id) throw `You are not the owner of the company ${req.params.id}, and cannot change it.`; //error on user, might have to declare on Company model
```
Comment (nlukic97) [inline ts/controllers/Company.controller.ts:88]: Error with TypeScript for the included User model for this query - still searching for solution

File: ts/controllers/Company.controller.ts [hunk 2]
```diff
@@ -187,14 +120,23 @@ const deleteCompany = async (req:Request, res:Response)=>{
...
                 id: req.params.id
-            }
+            },
+            include:User
         })
         
-        if(companyDelete){
-            return res.status(200).json({message:'Company with id '+ req.params.id + ' has been deleted.'})
+        if(company){
+            if(company.User.id !== req.user.id) throw `You are not the owner of the company ${req.params.id}, and cannot delete it.`; //error on user, might have to declare on Company model
```
Comment (nlukic97) [inline ts/controllers/Company.controller.ts:131]: Error with TypeScript for the included User model for this query - still searching for solution

File: ts/controllers/Profile.controller.ts [hunk 3]
```diff
@@ -111,122 +72,42 @@ const insertNewProfile = async (req: Request, res: Response) => {
...
+                id: req.params.id
+            },
+            include:User
+        })
+        
+        if(profileExists === null) throw `A profile with the id ${req.params.id} does not exist and cannot be assigned to this profile.`;
 
-        // checking if a company with a specific id exists (if it is supplied by the user in the req.body.user parameter)
+        // If the user who has the profile is not the same as the user who is making this request (from passport-jwt)
+        if(profileExists.User.id !== req.user.id) throw `You are not the owner of the profile ${req.params.id}, and cannot change it.`; //error on user, might have to declare on Company model
```
Comment (nlukic97) [inline ts/controllers/Profile.controller.ts:97]: Error with TypeScript for the included User model for this query - still searching for solution

File: ts/controllers/Profile.controller.ts [hunk 4]
```diff
@@ -248,20 +129,29 @@ const updateProfile = async (req: Request, res: Response) => {
...
                 id: req.params.id
-            }
+            },
+            include:User
         })
         
-        if(profileDelete){
-            return res.status(200).json({message:'Profile with id '+ req.params.id + ' has been deleted.'})
+        if(profile){
+            if(profile.User.id !== req.user.id) throw `You are not the owner of the profile ${req.params.id}, and cannot delete it.`; //error on user, might have to declare on Profile model
```
Comment (nlukic97) [inline ts/controllers/Profile.controller.ts:146]: 4. Error with TypeScript for the included User model for this query - still searching for solution

**Comment by nlukic97** [comment]
Update: Fxed the issue highlighted in the previous two comments with this commit: 
https://github.com/nlukic97/uteam-api/pull/6/commits/0693b75f266e863333d391e3450198573e3d96fe

In short, I added the **User** attribute to the **Profile** and **Company** interfaces, and assigned **UserAttributes** (the interface for User i imported) as the value for **User**.

File: ts/ts_interfaces/User.Interface.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export interface UserAttributes {
+    id?: number, 
+    username: string,
+    role?:'company-user'|'company-admin'|'superadmin', //optional since we have a default value for enum
```
Comment (vuxor) [inline ts/ts_interfaces/User.Interface.ts:4]: You can use enum instead of strings here

**Review by vuxor — CHANGES_REQUESTED** [review]

File: ts/middleware/validation/Company.update.validation.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    if(
+       (submitData.name && typeof(submitData.name) !== 'string')
+    || (submitData.logo && typeof(submitData.logo) !== 'string')
+    ) {
+        return res.status(400).json({message:'Please make sure the submitted parameters are of the correct type.'})
+    }
+
+    // trimming the available inputs, and returning an error if they are an empty string after trimming
+    if(submitData.name){
+        submitData.name = submitData.name.trim()
```
[the remark this answers is not in the document]
Comment (nlukic97) [reply ts/middleware/validation/Company.update.validation.ts]: This commit should fix this: https://github.com/nlukic97/uteam-api/pull/6/commits/dbd60a51f47af4949aacdaa345a50eb173fa427c

File: ts/middleware/validation/Profile.update.validation.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        if(req.body.profilePhoto === ''){
+            return res.status(400).json({message:'There appear to be empty fields. Please check your inputs and try again.'})
+        } else {
+            submitData.profilePhoto = req.body.profilePhoto
+        }
+    }
+
+    if(req.body.company !== undefined){        
+        if(req.body.company === ''){            
+            return res.status(400).json({message:'There appear to be empty fields. Please check your inputs and try again.'})
```
[the remark this answers is not in the document]
Comment (nlukic97) [reply ts/middleware/validation/Profile.update.validation.ts]: Fixed with this commit: https://github.com/nlukic97/uteam-api/commit/0c8250d3aada8479169d038636b6a586f30ea401

The code has been refactored to: 
- be more readable
- contain the specific error messages
- not contain unecessary nullchecks before trimming strings (same issue was present here as was in the validation for updating Company)

Sample: https://github.com/nlukic97/uteam-api/blob/0c8250d3aada8479169d038636b6a586f30ea401/ts/middleware/validation/Profile.update.validation.ts#L23-L25

File: ts/middleware/validation/Profile.update.validation.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    if(!submitData.name && !submitData.profilePhoto && !submitData.company){
+        return res.status(400).json({message:'No data submitted for update. Please submit the fields you would like to change (name, profilePhoto, and / or company).'})
+    }
+    
+    
+    //checking that present submitData is of correct type
+    if(
+       (submitData.name && typeof(submitData.name) !== 'string') 
+    || (submitData.profilePhoto && typeof(submitData.profilePhoto) !== 'string')
+    || (submitData.company && (Number.isInteger(submitData.company) === false || submitData.company === 0))
```
[the remark this answers is not in the document]
Comment (nlukic97) [reply ts/middleware/validation/Profile.update.validation.ts]: Fixed with the commit mentioned in the previous comment (https://github.com/nlukic97/uteam-api/commit/0c8250d3aada8479169d038636b6a586f30ea401)

File: ts/ts_interfaces/User.Interface.ts (hunk 5 above)
Comment (nlukic97) [reply ts/ts_interfaces/User.Interface.ts]: Managed to refactor using **enum**: https://github.com/nlukic97/uteam-api/commit/96e03ec1f146964f2c1d5e47c8194d3b9ad8c3ab

**Review by vuxor — APPROVED** [review]

**Pull request merged**